### PR TITLE
Protecting TypeScript library files

### PR DESCRIPTION
**Pull request by alexsaveliev**
- excluding them from regular files and putting to a separate object. This is done to prevent attempts to re-fetch them (by build server for example). This may lead to unwanted promise rejection and lack of hover/definition information. Addresses sourcegraph/sourcegraph#5192

### CODE REVIEW

File: src/project-manager.ts [hunk 1]
```diff
@@ -729,6 +730,25 @@ export class InMemoryFileSystem implements ts.ParseConfigHost, ts.ModuleResoluti
...
+	}
+
+	isTypeScriptLibrary(path: string): boolean {
+		return this.typeScriptLibraries.has(util.toUnixPath(path));
+	}
+
+	/**
+	 * Collects TypeScript library files
+	 */
+	private initTypeScriptLibraries() {
```
Comment (felixfbecker) [inline src/project-manager.ts]: This is doing sync IO when the InMemoryFileSystem is created, which is done for every new connection and will then block anything happening in other connections. Before it was done on startup, which was better.

File: src/project-manager.ts [hunk 2]
```diff
@@ -698,6 +694,11 @@ export interface FileSystemNode {
 export class InMemoryFileSystem implements ts.ParseConfigHost, ts.ModuleResolutionHost {
 
 	/**
+	 * Contains in-memory map (path => content) for all TypeScript libraries
+	 */
+	private typeScriptLibraries = new Map<string, string>();
```
Comment (felixfbecker) [inline src/project-manager.ts]: Idea: could we maybe just save the typescript libraries in the normal `files` Map, but with a custom URI scheme like `tslib:`? That's how I did it in PHP: https://sourcegraph.com/github.com/felixfbecker/php-language-server/-/blob/src/ComposerScripts.php#L55:1-56:1

**Comment by felixfbecker** [comment]
Should also have a test (can be a unit test)

File: src/project-manager.ts (hunk 2 above)
Comment (alexsaveliev) [inline src/project-manager.ts]: I think that keeping them in a separate object is safer :smile:

File: src/memfs.ts [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+/**
+ * TypeScript library files fetched from the local file system (bundled TS)
+ */
+let tsLibraries: Map<string, string>;
+
+/**
+ * Fetches TypeScript library files from local file system
+ */
+export function getTypeScriptLibraries(): Map<string, string> {
```
Comment (felixfbecker) [inline src/memfs.ts]: Instead of this getter function, lets just use the map directly. Does it even have to be exported?

File: src/memfs.ts [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+	 */
+	rootNode: FileSystemNode;
+
+	constructor(path: string, private logger: Logger = new NoopLogger()) {
+		this.path = path;
+		this.overlay = new Map<string, string>();
+		this.rootNode = { file: false, children: new Map<string, FileSystemNode>() };
+	}
+
+	isTypeScriptLibrary(path: string): boolean {
```
Comment (felixfbecker) [inline src/memfs.ts]: I don't see a reason for this to be a method of InMemorySystem now

Comment (alexsaveliev) [inline src/memfs.ts]: fixed

File: src/memfs.ts (hunk 3 above)
Comment (alexsaveliev) [inline src/memfs.ts]: exposing map directly. It's still needs to be exported for tests

**Pull request merged**